game/video.cc

```diff
@@ -1,27 +1,101 @@
 #include "video.hh"
 
+#include "ffmpeg.hh"
 #include "util.hh"
 #include <cmath>
 
-Video::Video(fs::path const& _videoFile, double videoGap): m_mpeg(_videoFile), m_videoGap(videoGap), m_textureTime(), m_lastTime(), m_alpha(-0.5, 1.5) {}
+bool Video::tryPop(Bitmap& f, double timestamp) {
+	std::unique_lock<std::mutex> l(m_mutex);
+
+	// if timestamp is out of the queue's range, ask a seek
+	// FIXME 4 should be linked to queue depth: we know the frame rate, thus how
+	// many frames needed for an arbitrary duration
+	m_seek_asked |= timestamp > m_readPosition + 4 || timestamp < m_readPosition;
+
+	m_readPosition = timestamp;
+
+	// if queue is not empty, we are sure will remove at least one element from it.
+	// When seeking needed to wake up thread if waiting for rooms.
+	if (!m_queue.empty() || m_seek_asked) m_cond.notify_all();
+
+	if (m_seek_asked) return false;
+
+	// discard outdated frames retaining only the most recent frame that is _before_ timestamp
+	while (!m_queue.empty() && std::next(m_queue.begin()) != m_queue.end() && std::next(m_queue.begin())->timestamp < timestamp) m_queue.pop_front();
+
+	if (m_queue.empty() || m_queue.front().timestamp > timestamp) return false; // Nothing to deliver
+
+	f = std::move(m_queue.front());
+	m_queue.pop_front();
+	return true;
+}
+
+void Video::push(Bitmap&& f) {
+	std::unique_lock<std::mutex> l(m_mutex);
+	m_cond.wait(l, [this]{ return m_quit || m_seek_asked || m_queue.size() < m_max; });
+	if (m_quit || m_seek_asked) return; // Drop frame when seek/quit asked
+	m_queue.emplace_back(std::move(f));
+}
+
+Video::~Video() { 
+	{
+		std::lock_guard<std::mutex> l(m_mutex);
+		m_quit = true;
+	}
+	m_cond.notify_all();
+	m_grabber.get();
+}
+
+Video::Video(fs::path const& _videoFile, double videoGap): m_videoGap(videoGap), m_textureTime(), m_alpha(-0.5, 1.5) {
+	// make ffmpeg here to get any exception in the current thread
+	auto ffmpeg = std::make_unique<VideoFFmpeg>(_videoFile, [this] (auto f) { push(std::move(f)); });
+	m_grabber = std::async(std::launch::async, [this, file = _videoFile, ffmpeg = std::move(ffmpeg)] {
+			int errors = 0;
+			std::unique_lock<std::mutex> l(m_mutex);
```

Comment:
Two levels of indent? I don't think this is covered in the style guide but isn't the normal practice to indent only once even if multiple braces are opened on the same line?

Reply:
done